### PR TITLE
feat: created adding payment method flow

### PR DESCRIPTION
**Pull request by vackca**

### CODE REVIEW

**Comment by vackca** [comment]
I am gonna add tests in the next PR

File: microservices/payment/src/interfaces/stripe-payment-options.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import Stripe from 'stripe';
+import StripePaymentMethods from '@constants/stripe-payment-methods';
```
Comment (MatthewPattell) [inline microservices/payment/src/interfaces/stripe-payment-options.ts]:
```suggestion
import type StripePaymentMethods from '@constants/stripe-payment-methods';
```

File: microservices/payment/src/interfaces/stripe-payment-options.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+import Stripe from 'stripe';
```
Comment (MatthewPattell) [inline microservices/payment/src/interfaces/stripe-payment-options.ts]:
```suggestion
import type Stripe from 'stripe';
```

File: microservices/payment/src/interfaces/stripe-payment-options.ts [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+import StripePaymentMethods from '@constants/stripe-payment-methods';
+
+export default interface IStripePaymentOptions {
```
Comment (MatthewPattell) [inline microservices/payment/src/interfaces/stripe-payment-options.ts]:
```suggestion
export default interface IStripeOptions {
```

File: microservices/payment/src/methods/customer/create.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
+import { Endpoint } from '@lomray/microservice-helpers';
+import { Type } from 'class-transformer';
+import { IsObject } from 'class-validator';
+import { getManager } from 'typeorm';
+import Customer from '@entities/customer';
+import { ICustomerParams } from '@services/payment-gateway/abstract';
+import Factory from '@services/payment-gateway/factory';
+
+class CustomerCreateInput implements ICustomerParams {
+  userId: string;
```
Comment (MatthewPattell) [inline microservices/payment/src/methods/customer/create.ts:10]: add decorator

File: microservices/payment/src/methods/customer/create.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
+import { Endpoint } from '@lomray/microservice-helpers';
+import { Type } from 'class-transformer';
+import { IsObject } from 'class-validator';
+import { getManager } from 'typeorm';
+import Customer from '@entities/customer';
+import { ICustomerParams } from '@services/payment-gateway/abstract';
```
Comment (MatthewPattell) [inline microservices/payment/src/methods/customer/create.ts]:
```suggestion
import type { ICustomerParams } from '@services/payment-gateway/abstract';
```

File: microservices/payment/src/services/payment-gateway/stripe.ts [hunk 6]
```diff
@@ -10,16 +11,49 @@ class Stripe extends Abstract {
...
+      payment_method_types: this.paymentOptions.methods,
+    });
+
+    return clientSecret;
+  }
+
+  /**
+   * Create Customer entity
+   */
+  async createCustomer(userId: string): Promise<Customer> {
```
Comment (MatthewPattell) [inline microservices/payment/src/services/payment-gateway/stripe.ts]:
```suggestion
 public async createCustomer(userId: string): Promise<Customer> {
```

File: microservices/payment/src/services/payment-gateway/abstract.ts [hunk 7]
```diff
@@ -10,6 +14,16 @@ export interface IBankAccountParams {
   test: boolean;
 }
 
+export interface ICustomerParams {
+  userId: string;
+}
+
+export interface ISetupIntentParams {
+  userId: string;
+}
```
Comment (MatthewPattell) [inline microservices/payment/src/services/payment-gateway/abstract.ts]:
```suggestion
```

File: microservices/payment/src/services/payment-gateway/abstract.ts [hunk 8]
```diff
@@ -10,6 +14,16 @@ export interface IBankAccountParams {
...
+
+export type TPaymentEntity = Stripe;
```
Comment (MatthewPattell) [inline microservices/payment/src/services/payment-gateway/abstract.ts]:
```suggestion
```

File: microservices/payment/src/services/payment-gateway/abstract.ts [hunk 9]
```diff
@@ -22,17 +36,31 @@ abstract class Abstract {
...
    * @protected
    */
-  protected readonly paymentOptions: Record<string, any>;
+  protected readonly paymentOptions: TPaymentOptions;
+
+  /**
+   * @protected
+   */
+  protected readonly paymentEntity: TPaymentEntity;
+
```
Comment (MatthewPattell) [inline microservices/payment/src/services/payment-gateway/abstract.ts]:
```suggestion
```

File: microservices/payment/src/services/payment-gateway/abstract.ts [hunk 10]
```diff
@@ -22,17 +36,31 @@ abstract class Abstract {
...
 
   /**
    * @constructor
    */
   public constructor(
     paymentProvider: Abstract['paymentProvider'],
-    paymentOptions: Abstract['paymentOptions'] = {},
+    paymentEntity: Abstract['paymentEntity'],
+    manager: EntityManager,
+    paymentOptions: Abstract['paymentOptions'],
```
Comment (MatthewPattell) [inline microservices/payment/src/services/payment-gateway/abstract.ts]:
```suggestion
    manager: EntityManager,
```

File: microservices/payment/src/services/payment-gateway/abstract.ts [hunk 11]
```diff
@@ -44,6 +72,16 @@ abstract class Abstract {
...
+
+  /**
+   * Create new customer
+   */
+  public abstract createCustomer(userId: string): Promise<Customer>;
+
+  /**
+   * Create setup intent
+   */
+  public abstract createSetupIntent(userId: string): Promise<string | null>;
```
Comment (MatthewPattell) [inline microservices/payment/src/services/payment-gateway/abstract.ts]:
```suggestion
```

File: microservices/payment/src/services/payment-gateway/factory.ts [hunk 12]
```diff
@@ -10,12 +12,18 @@ class Factory {
...
     switch (paymentProvider) {
       case PaymentProvider.STRIPE:
-        return new Stripe(paymentProvider, paymentOptions);
+        if (paymentOptions) {
+          const stripe = new Stripe(paymentOptions.apiKey, paymentOptions.config);
+
+          return new StripeEntity(paymentProvider, stripe, manager, paymentOptions);
+        }
+
+        throw new Error(`Payment options for stripe are not provided`);
```
Comment (MatthewPattell) [inline microservices/payment/src/services/payment-gateway/factory.ts]:
```suggestion
        if (!paymentOptions) {
            throw new Error('Payment options for stripe are not provided');
        }

         return new Stripe(manager, paymentProvider, paymentOptions);
```

File: microservices/payment/src/services/payment-gateway/factory.ts [hunk 13]
```diff
@@ -1,7 +1,9 @@
+import Stripe from 'stripe';
+import { EntityManager } from 'typeorm';
 import remoteConfig from '@config/remote';
 import PaymentProvider from '@constants/payment-provider';
 import Abstract from './abstract';
-import Stripe from './stripe';
+import StripeEntity from './stripe';
```
Comment (MatthewPattell) [inline microservices/payment/src/services/payment-gateway/factory.ts]:
```suggestion
import Stripe from './stripe';
```

File: microservices/payment/src/services/payment-gateway/factory.ts [hunk 14]
```diff
@@ -1,7 +1,9 @@
+import Stripe from 'stripe';
+import { EntityManager } from 'typeorm';
```
Comment (MatthewPattell) [inline microservices/payment/src/services/payment-gateway/factory.ts]:
```suggestion
import type { EntityManager } from 'typeorm';
```

File: microservices/payment/src/services/payment-gateway/stripe.ts [hunk 15]
```diff
@@ -1,7 +1,8 @@
+import StripeEntity from 'stripe';
```
Comment (MatthewPattell) [inline microservices/payment/src/services/payment-gateway/stripe.ts]:
```suggestion
import StripeSDK from 'stripe';
```

File: microservices/payment/src/methods/index.ts [hunk 16]
```diff
@@ -21,6 +24,13 @@ export default (ms: Microservice): void => {
       add: BankAccountAdd,
     },
     transaction: CrudTransaction,
+    customer: {
+      ...CrudCustomer,
+      create: CustomerCreate,
+    },
+    setupIntent: {
+      create: CreateSetupIntent,
+    },
```
Comment (MatthewPattell) [inline microservices/payment/src/methods/index.ts]:
```suggestion
```

File: microservices/payment/src/methods/customer/create.ts [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+import Factory from '@services/payment-gateway/factory';
+
+class CustomerCreateInput implements ICustomerParams {
```
Comment (MatthewPattell) [inline microservices/payment/src/methods/customer/create.ts]:
```suggestion
class CustomerCreateInput {
```

File: microservices/payment/src/methods/setup-intent/create.ts [hunk 18]
```diff
@@ -0,0 +1,32 @@
+import { Endpoint } from '@lomray/microservice-helpers';
+import { getManager } from 'typeorm';
+import { ISetupIntentParams } from '@services/payment-gateway/abstract';
+import Factory from '@services/payment-gateway/factory';
+
+class CreateSetupIntentInput implements ISetupIntentParams {
```
Comment (MatthewPattell) [inline microservices/payment/src/methods/setup-intent/create.ts]:
```suggestion
class CreateSetupIntentInput {
```

File: microservices/payment/src/methods/setup-intent/create.ts [hunk 19]
```diff
@@ -0,0 +1,32 @@
+import { Endpoint } from '@lomray/microservice-helpers';
```
Comment (MatthewPattell) [inline microservices/payment/src/methods/setup-intent/create.ts]: microservices/payment/src/methods/setup-intent/create.ts => microservices/payment/src/methods/stripe/setup-intent.ts

File: microservices/payment/src/services/payment-gateway/stripe.ts [hunk 20]
```diff
@@ -10,16 +11,49 @@ class Stripe extends Abstract {
...
    */
-  addBankAccount(params: IBankAccountParams): Promise<BankAccount> {
+  addBankAccount(): Promise<BankAccount> {
     return Promise.resolve(new BankAccount());
   }
+
+  /**
+   * Create SetupIntent and get back client secret
+   */
+  async createSetupIntent(userId: string): Promise<string | null> {
```
Comment (MatthewPattell) [inline microservices/payment/src/services/payment-gateway/stripe.ts]:
```suggestion
public  async setupIntent(userId: string): Promise<string | null> {
```

File: microservices/payment/src/methods/setup-intent/create.ts [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+ * Create setupIntent and return client token to get access to adding payment method
+ */
+const createSetupIntent = Endpoint.custom(
+  () => ({
+    input: CreateSetupIntentInput,
+    output: CreateSetupIntentOutput,
+    description: 'Creates setup intent and return client secret key',
+  }),
+  async ({ userId }) => {
+    const service = await Factory.create(getManager());
```
Comment (MatthewPattell) [inline microservices/payment/src/methods/setup-intent/create.ts]:
```suggestion
    const service = await Factory.create(getManager()) as Stripe;
```

**Review by MatthewPattell — CHANGES_REQUESTED** [review]

File: microservices/payment/src/interfaces/stripe-payment-options.ts (hunk 1 above)
Comment (vackca) [reply microservices/payment/src/interfaces/stripe-payment-options.ts]: added

File: microservices/payment/src/interfaces/stripe-payment-options.ts (hunk 2 above)
Comment (vackca) [reply microservices/payment/src/interfaces/stripe-payment-options.ts]: added

File: microservices/payment/src/interfaces/stripe-payment-options.ts (hunk 3 above)
Comment (vackca) [reply microservices/payment/src/interfaces/stripe-payment-options.ts]: changed

File: microservices/payment/src/methods/customer/create.ts (hunk 5 above)
Comment (vackca) [reply microservices/payment/src/methods/customer/create.ts]: added

File: microservices/payment/src/methods/index.ts (hunk 16 above)
Comment (vackca) [reply microservices/payment/src/methods/index.ts]: deleted

File: microservices/payment/src/methods/customer/create.ts (hunk 17 above)
Comment (vackca) [reply microservices/payment/src/methods/customer/create.ts]: changed

File: microservices/payment/src/methods/setup-intent/create.ts (hunk 18 above)
Comment (vackca) [reply microservices/payment/src/methods/setup-intent/create.ts]: deleted

File: microservices/payment/src/methods/setup-intent/create.ts (hunk 19 above)
Comment (vackca) [reply microservices/payment/src/methods/setup-intent/create.ts]: renamed

File: microservices/payment/src/services/payment-gateway/stripe.ts (hunk 20 above)
Comment (vackca) [reply microservices/payment/src/services/payment-gateway/stripe.ts]: added

File: microservices/payment/src/methods/setup-intent/create.ts (hunk 21 above)
Comment (vackca) [reply microservices/payment/src/methods/setup-intent/create.ts]: added

File: microservices/payment/src/services/payment-gateway/stripe.ts (hunk 15 above)
Comment (vackca) [reply microservices/payment/src/services/payment-gateway/stripe.ts]: renamed

File: microservices/payment/src/services/payment-gateway/abstract.ts (hunk 11 above)
Comment (vackca) [reply microservices/payment/src/services/payment-gateway/abstract.ts]: deleted

File: microservices/payment/src/services/payment-gateway/factory.ts (hunk 12 above)
Comment (vackca) [reply microservices/payment/src/services/payment-gateway/factory.ts]: changed

File: microservices/payment/src/services/payment-gateway/factory.ts (hunk 13 above)
Comment (vackca) [reply microservices/payment/src/services/payment-gateway/factory.ts]: changed

File: microservices/payment/src/services/payment-gateway/factory.ts (hunk 14 above)
Comment (vackca) [reply microservices/payment/src/services/payment-gateway/factory.ts]: added

File: microservices/payment/src/services/payment-gateway/abstract.ts (hunk 10 above)
Comment (vackca) [reply microservices/payment/src/services/payment-gateway/abstract.ts]: deleted

File: microservices/payment/src/services/payment-gateway/abstract.ts (hunk 9 above)
Comment (vackca) [reply microservices/payment/src/services/payment-gateway/abstract.ts]: deleted

File: microservices/payment/src/services/payment-gateway/abstract.ts (hunk 8 above)
Comment (vackca) [reply microservices/payment/src/services/payment-gateway/abstract.ts]: deleted

File: microservices/payment/src/services/payment-gateway/stripe.ts (hunk 6 above)
Comment (vackca) [reply microservices/payment/src/services/payment-gateway/stripe.ts]: added

File: microservices/payment/src/services/payment-gateway/abstract.ts (hunk 7 above)
Comment (vackca) [reply microservices/payment/src/services/payment-gateway/abstract.ts]: deleted

File: microservices/payment/src/methods/customer/create.ts (hunk 4 above)
Comment (vackca) [reply microservices/payment/src/methods/customer/create.ts:10]: added

File: microservices/payment/src/interfaces/remote-config.ts [hunk 22]
```diff
@@ -1,9 +1,10 @@
 import PaymentProvider from '@constants/payment-provider';
+import TPaymentOptions from '@interfaces/payment-options';
```
Comment (MatthewPattell) [inline microservices/payment/src/interfaces/remote-config.ts]:
```suggestion
import type TPaymentOptions from '@interfaces/payment-options';
```

File: microservices/payment/src/methods/stripe/setup-intent.ts [hunk 23]
```diff
@@ -0,0 +1,32 @@
...
+import Factory from '@services/payment-gateway/factory';
+import Stripe from '@services/payment-gateway/stripe';
+
+class SetupIntentInput {
+  userId: string;
+}
+
+class SetupIntentOutput {
+  clientSecretToken: string | null;
+}
```
Comment (MatthewPattell) [inline microservices/payment/src/methods/stripe/setup-intent.ts]: where validation decorators?

File: microservices/payment/src/methods/stripe/setup-intent.ts [hunk 24]
```diff
@@ -0,0 +1,32 @@
...
+ * Create setupIntent and return client token to get access to adding payment method
+ */
+const setupIntent = Endpoint.custom(
+  () => ({
+    input: SetupIntentInput,
+    output: SetupIntentOutput,
+    description: 'Creates setup intent and return client secret key',
+  }),
+  async ({ userId }) => {
+    const service = (await Factory.create(getManager())) as Stripe;
```
Comment (MatthewPattell) [inline microservices/payment/src/methods/stripe/setup-intent.ts:36]: we should add `if` and check payment provider === 'stripe'

File: microservices/payment/src/services/payment-gateway/factory.ts [hunk 25]
```diff
@@ -10,13 +11,16 @@ class Factory {
...
     const { paymentProvider, paymentOptions } = await remoteConfig();
 
     switch (paymentProvider) {
       case PaymentProvider.STRIPE:
-        return new Stripe(paymentProvider, paymentOptions);
+        if (!paymentOptions) {
+          throw new Error(`Payment options for stripe are not provided`);
+        }
 
+        return new Stripe(manager, paymentProvider, paymentOptions);
```
Comment (MatthewPattell) [inline microservices/payment/src/services/payment-gateway/factory.ts:23]:
```suggestion
        return new Stripe(manager, paymentProvider, paymentOptions);

```

File: microservices/payment/src/services/payment-gateway/stripe.ts [hunk 26]
```diff
@@ -1,25 +1,82 @@
+import StripeSdk from 'stripe';
+import { EntityManager } from 'typeorm';
```
Comment (MatthewPattell) [inline microservices/payment/src/services/payment-gateway/stripe.ts]:
```suggestion
import type { EntityManager } from 'typeorm';
```

File: microservices/payment/src/services/payment-gateway/stripe.ts [hunk 27]
```diff
@@ -1,25 +1,82 @@
...
+      payment_method_types: this.paymentOptions.methods,
+    });
+
+    return clientSecret;
+  }
+
+  /**
+   * Create Customer entity
+   */
+  public async createCustomer(userId: string): Promise<Customer> {
```
Comment (MatthewPattell) [inline microservices/payment/src/services/payment-gateway/stripe.ts:68]: detect existing customer??

File: microservices/payment/src/services/payment-gateway/stripe.ts [hunk 28]
```diff
@@ -1,25 +1,82 @@
...
+
+  /**
+   * @protected
+   */
+  protected readonly paymentEntity: StripeSdk;
+
+  constructor(
+    manager: EntityManager,
+    paymentProvider: Abstract['paymentProvider'],
+    paymentOptions: TPaymentOptions,
```
Comment (MatthewPattell) [inline microservices/payment/src/services/payment-gateway/stripe.ts]: use stripe interface directly?

**Review by MatthewPattell — CHANGES_REQUESTED** [review]
update permissions?

File: microservices/payment/src/interfaces/remote-config.ts (hunk 22 above)
Comment (vackca) [reply microservices/payment/src/interfaces/remote-config.ts]: added

File: microservices/payment/src/methods/stripe/setup-intent.ts (hunk 23 above)
Comment (vackca) [reply microservices/payment/src/methods/stripe/setup-intent.ts]: added

File: microservices/payment/src/services/payment-gateway/factory.ts (hunk 25 above)
Comment (vackca) [reply microservices/payment/src/services/payment-gateway/factory.ts:23]: added

File: microservices/payment/src/methods/stripe/setup-intent.ts (hunk 24 above)
Comment (vackca) [reply microservices/payment/src/methods/stripe/setup-intent.ts:36]: added

File: microservices/payment/src/services/payment-gateway/stripe.ts (hunk 26 above)
Comment (vackca) [reply microservices/payment/src/services/payment-gateway/stripe.ts]: added

File: microservices/payment/src/services/payment-gateway/stripe.ts (hunk 28 above)
Comment (vackca) [reply microservices/payment/src/services/payment-gateway/stripe.ts]: used

File: microservices/payment/src/services/payment-gateway/stripe.ts [hunk 29]
```diff
@@ -1,25 +1,87 @@
...
+
+    return clientSecret;
+  }
+
+  /**
+   * Create Customer entity
+   */
+  public async createCustomer(userId: string): Promise<Customer> {
+    if (await this.customerRepository.findOne(userId)) {
+      throw new Error('Customer for this user already exists');
```
Comment (MatthewPattell) [inline microservices/payment/src/services/payment-gateway/stripe.ts]:
```suggestion
      throw new Error('Customer already exists.');
```

File: microservices/payment/src/services/payment-gateway/stripe.ts [hunk 30]
```diff
@@ -1,25 +1,87 @@
...
+  protected readonly paymentEntity: StripeSdk;
+
+  constructor(
+    manager: EntityManager,
+    paymentProvider: PaymentProvider.STRIPE,
+    paymentOptions: TPaymentOptions,
+  ) {
+    super(paymentProvider, manager);
+
+    this.paymentOptions = paymentOptions;
```
Comment (MatthewPattell) [inline microservices/payment/src/services/payment-gateway/stripe.ts]: it should be in abstract

File: microservices/payment/src/services/payment-gateway/stripe.ts (hunk 29 above)
Comment (MatthewPattell) [inline microservices/payment/src/services/payment-gateway/stripe.ts]: return customer instead error?

File: microservices/payment/src/services/payment-gateway/stripe.ts [hunk 31]
```diff
@@ -1,25 +1,87 @@
...
+   * @protected
+   */
+  protected readonly paymentOptions: IStripeOptions;
+
+  /**
+   * @protected
+   */
+  protected readonly paymentEntity: StripeSdk;
+
+  constructor(
```
Comment (MatthewPattell) [inline microservices/payment/src/services/payment-gateway/stripe.ts:27]:
```suggestion
  /**
   * @constructor
   */
  constructor(
```

File: microservices/payment/src/services/payment-gateway/stripe.ts [hunk 32]
```diff
@@ -1,25 +1,87 @@
+import StripeSdk from 'stripe';
+import type { EntityManager } from 'typeorm';
+import type PaymentProvider from '@constants/payment-provider';
 import BankAccount from '@entities/bank-account';
 import Card from '@entities/card';
+import Customer from '@entities/customer';
+import type TPaymentOptions from '@interfaces/payment-options';
+import IStripeOptions from '@interfaces/stripe-options';
```
Comment (MatthewPattell) [inline microservices/payment/src/services/payment-gateway/stripe.ts]:
```suggestion
import type IStripeOptions from '@interfaces/stripe-options';
```

File: microservices/authorization/migrations/permissions/list/methods/payment.json [hunk 33]
```diff
@@ -0,0 +1,137 @@
...
+    "modelIn": "RemoveRequestParams",
+    "modelOut": "payment.RemoveOutputParams.87d10388e62bc5cd27e433c97b2b7ea5",
+    "methodFilters": []
+  },
+  {
+    "microservice": "payment",
+    "method": "customer.create",
+    "description": "Create new customer",
+    "allowGroup": [
+      "user"
```
Comment (MatthewPattell) [inline microservices/authorization/migrations/permissions/list/methods/payment.json]:
```suggestion
      "admin"
```

File: microservices/payment/src/services/payment-gateway/stripe.ts (hunk 30 above)
Comment (MatthewPattell) [inline microservices/payment/src/services/payment-gateway/stripe.ts]:
```suggestion
```

File: microservices/payment/src/services/payment-gateway/stripe.ts [hunk 34]
```diff
@@ -1,25 +1,87 @@
...
+
+    const { client_secret: clientSecret } = await this.paymentEntity.setupIntents.create({
+      customer: customer.customerId,
+      // eslint-disable-next-line camelcase
+      payment_method_types: this.paymentOptions.methods,
+    });
+
+    return clientSecret;
+  }
+
```
Comment (MatthewPattell) [inline microservices/payment/src/services/payment-gateway/stripe.ts:64]: protected getCustomer()

File: microservices/payment/src/services/payment-gateway/stripe.ts (hunk 29 above)
Comment (MatthewPattell) [reply microservices/payment/src/services/payment-gateway/stripe.ts]:
```
public async createCustomer(userId: string): Promise<Customer> {
   const stripeCustomer: StripeSdk.Customer = await this.paymentEntity.customers.create();
  return super.createCustomer(userId, stripeCustomer.id);
}
```

File: microservices/payment/src/services/payment-gateway/abstract.ts [hunk 35]
```diff
@@ -44,6 +43,11 @@ abstract class Abstract {
    * Add new bank account
    */
   public abstract addBankAccount(params: IBankAccountParams): Promise<BankAccount>;
+
```
Comment (MatthewPattell) [inline microservices/payment/src/services/payment-gateway/abstract.ts:58]:
```
protected getCustomer() {
  const customer = await findOne();

  if (customer) {
      return customer;
  }

  const newCustomer = await this.createCustomer(usertId);

  return newCustomer;
}

public createCustomer(userId: string, customerId?: =uuidv4()) {
   const customer =  this.customerRepository.create({
      customerId,
      userId,
    });

  await repo.save(newCustomer)

 return customer;
}
```

**Review by MatthewPattell — CHANGES_REQUESTED** [review]

File: microservices/payment/src/services/payment-gateway/abstract.ts [hunk 36]
```diff
@@ -22,17 +27,24 @@ abstract class Abstract {
   /**
    * @protected
    */
-  protected readonly paymentOptions: Record<string, any>;
+  protected readonly customerRepository: Repository<Customer>;
+
+  /**
+   * @protected
+   */
+  protected readonly paymentOptions: IStripeOptions;
```
Comment (MatthewPattell) [inline microservices/payment/src/services/payment-gateway/abstract.ts]:
```suggestion
  protected readonly paymentOptions: TPaymentOptions;
```

**Review by MatthewPattell — APPROVED** [review]

**Review by MatthewPattell — APPROVED** [review]

**Pull request merged**